src/util.rs

```diff
@@ -7,3 +7,82 @@ macro_rules! regex {
         &REGEX
     }};
 }
+
+use crate::char_len;
+use std::{collections::HashMap, iter::FromIterator};
+
+pub struct TranslationTable {
+    table: HashMap<char, char>,
+}
+impl TranslationTable {
+    pub fn new(f: &str, t: &str) -> Result<Self, &'static str> {
+        if char_len(f) != char_len(t) {
+            Err("Arguments passed to `TranslationTable::new` must have equal character length")
+        } else {
+            Ok(TranslationTable {
+                table: HashMap::from_iter(f.chars().zip(t.chars())),
+            })
+        }
+    }
+
+    pub fn translate(&self, s: String) -> String {
```

Comment:
Should be able to take in a `&str` as input to avoid having to clone where we use it, no?
```suggestion
    pub fn translate(&self, s: &str) -> String {
```